### PR TITLE
Toilet and dynamic pads

### PR DESCRIPTION
**Pull request by Hajto**

### CODE REVIEW

File: chapter1/pipeline.md [hunk 1]
```diff
@@ -142,3 +142,37 @@ Pipeline.play(pid)
...
+If you want to connect two pads, namely a pad that is in `push` mode to one that 
+is in `pull` mode, you MUST use a toilet. The toilet is a container that
+stores buffer send by `push` element and allows `pull` element to
+place demands. Creating a toilet is quite a simple feat. 
+When linking pads just use the third slot of the tuple.
+
+```elixir
+links = %{
+  #...
+  {:push, :source} => {:pull, :sink, toilet: true},
```
Comment (bblaszkow06) [inline chapter1/pipeline.md]: Pads should be called input/output

File: chapter1/pipeline.md [hunk 2]
```diff
@@ -142,3 +142,37 @@ Pipeline.play(pid)
...
+
+You may ask a question, what to do if the toilet keeps overflowing?
+Create a bigger toilet of course... Toilet is more flexible than you might think
+you can not only enable it but also configure its size.
+
+```elixir
+toilet_config = %{warn: warn_threshold, fail: fail_threshold}
+links = %{
+  #...
+  {:push, :source} => {:pull, :sink, toilet: toilet_config},
```
Comment (bblaszkow06) [inline chapter1/pipeline.md]: Pads should be called input/output

File: chapter1/pipeline.md [hunk 3]
```diff
@@ -142,3 +142,37 @@ Pipeline.play(pid)
...
 
 The given `.mp3` file should be played on default device in your system.
+
+## Connecting `push` pad to `pull` pad
+
+If you want to connect two pads, namely a pad that is in `push` mode to one that 
+is in `pull` mode, you MUST use a toilet. The toilet is a container that
+stores buffer send by `push` element and allows `pull` element to
+place demands. Creating a toilet is quite a simple feat. 
+When linking pads just use the third slot of the tuple.
```
Comment (bblaszkow06) [inline chapter1/pipeline.md]: I'm not sure we can speak about slot here, it suggests that slot was always there while actually, we changed 2-tuple into 3-tuple. I would rephrase this sentence into just: `When linking pads, just add info about toilet to a proper input pad`.

File: chapter2/dynamic_pads.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# Dynamic Pads
+
+Dynamic pads, are pads that are created on demand when linking elements.
+
+## Why do we need dynamic pads?
+
+In some applications manually specifying pads isn't good enough.
+Let's consider sound mixer. If we are talking about physical piece
```
Comment (bblaszkow06) [inline chapter2/dynamic_pads.md]: `Piece` seems off to me. Did you mean physical device?

File: chapter2/dynamic_pads.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+it will usually have 8, 16, 24 or even in some extreme cases 768 inputs.
+Writing specifications for all those pads would be a hassle, wouldn't it?
+
+## Creating an element with dynamic pads
+
+Creating an element with dynamic pads not much different than is not much
```
Comment (bblaszkow06) [inline chapter2/dynamic_pads.md]: ~not much different than~

File: chapter2/dynamic_pads.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+key to `:on_request`.
+
+```elixir
+def_input_pads pad_name [
+  availability: :on_request, 
+  # other properties
+  ]
+```
+
+Now when handling incoming buffers pads will be suffixed with subsequent numbers.
```
Comment (bblaszkow06) [inline chapter2/dynamic_pads.md]: This is not precise and misleading. Suggestion:
```
Now, each time some element is linked to this pad, a new instance of the pad is created and callback 
`handle_pad_added` (tu poprawić, że by był link) is invoked. Instances of a pad can be referenced
as `{:dynamic, :input, number}`
```

File: chapter2/dynamic_pads.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+## Handling events
+
+There is one problem though? What if Event such as End of Stream is passed
```
Comment (bblaszkow06) [inline chapter2/dynamic_pads.md]: That question mark seems off.

File: chapter2/dynamic_pads.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  # other properties
+  ]
+```
+
+Now when handling incoming buffers pads will be suffixed with subsequent numbers.
+
+## Handling events
+
+There is one problem though? What if Event such as End of Stream is passed 
+through a pad? Usually if you are using pads `:input` and `:output` default
```
Comment (bblaszkow06) [inline chapter2/dynamic_pads.md]: the default

File: chapter2/dynamic_pads.md [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+  ]
+```
+
+Now when handling incoming buffers pads will be suffixed with subsequent numbers.
+
+## Handling events
+
+There is one problem though? What if Event such as End of Stream is passed 
+through a pad? Usually if you are using pads `:input` and `:output` default 
+action is forwarding event. Forwarding means that if event comes to
```
Comment (bblaszkow06) [inline chapter2/dynamic_pads.md]: is to forward an event. It means that if an event comes at `:input` pad it is set via `:output` pad and vice versa.

File: chapter2/dynamic_pads.md [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+```elixir
+def_input_pads pad_name [
+  availability: :on_request, 
+  # other properties
+  ]
+```
+
+Now when handling incoming buffers pads will be suffixed with subsequent numbers.
+
+## Handling events
```
Comment (bblaszkow06) [inline chapter2/dynamic_pads.md]: This paragraph should mention somewhere it only applies to filters

File: chapter2/dynamic_pads.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+There is one problem though? What if Event such as End of Stream is passed 
+through a pad? Usually if you are using pads `:input` and `:output` default 
+action is forwarding event. Forwarding means that if event comes to
+`:input` pad it is sent to `:output` pad and vice versa.
+
+There is one problem though, which of dynamic pad would be considered `:input`
```
Comment (bblaszkow06) [inline chapter2/dynamic_pads.md]: `There is one problem though` is repeated

File: chapter2/dynamic_pads.md (hunk 4 above)
Comment (Hajto) [reply chapter2/dynamic_pads.md]: https://www.merriam-webster.com/dictionary/piece
11th meaning.

Yup I meant physical device.
